### PR TITLE
Docker folder for Magenta library repo

### PR DESCRIPTION
**Pull request by werkaaa**
Docker folder that will be added to the Magenta library.

- Simple Dockerfile with Github installation of the library
- Sample config file
- task.py (wrapping script)
- tests for task.py
- Documentation

### CODE REVIEW

File: magenta_docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Uses one of AI Platform base images.
+# You can try using different images however only this one has been tested.
+FROM gcr.io/deeplearning-platform-release/tf2-gpu.2-2
+WORKDIR /root
```
Comment (marmolejo) [inline magenta_docker/Dockerfile]: Move this close where it's needed.

File: magenta_docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+# Installs sndfile library for reading and writing audio files.
+RUN apt-get update && apt-get install -y libsndfile-dev
+
+# Upgrades Tensorflow and Tensorflow Probability
+# Newer version of Tensorflow is needed for multiple VMs training
+RUN pip install --upgrade pip && pip install --upgrade tensorflow tensorflow-probability
```
Comment (marmolejo) [inline magenta_docker/Dockerfile]: You can split it in 2 lines by using a backslash:
```
RUN pip install --upgrade pip && \
  pip install --upgrade tensorflow tensorflow-probability
```

File: magenta_docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# Uses one of AI Platform base images.
+# You can try using different images however only this one has been tested.
+FROM gcr.io/deeplearning-platform-release/tf2-gpu.2-2
+WORKDIR /root
+
+# Installs sndfile library for reading and writing audio files.
+RUN apt-get update && apt-get install -y libsndfile-dev
```
Comment (marmolejo) [inline magenta_docker/Dockerfile]: Use `--no-install-recommends` too to avoid installing unintended packages.

File: magenta_docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+# Upgrades Tensorflow and Tensorflow Probability
+# Newer version of Tensorflow is needed for multiple VMs training
+RUN pip install --upgrade pip && pip install --upgrade tensorflow tensorflow-probability
+
+# Installs Magenta DDSP from Github.
+RUN wget https://github.com//magenta/ddsp/archive/master.zip
+RUN unzip master.zip
+RUN cd ddsp-master && python setup.py install
```
Comment (marmolejo) [inline magenta_docker/Dockerfile]: I believe we want all these commands in a single RUN so the fs checkpoint doesn't store temporary files:
```
RUN wget https://github.com//magenta/ddsp/archive/master.zip && \
  unzip master.zip && \
  cd ddsp-master && \
  python setup.py install
```

File: magenta_docker/README.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+* Install [Docker](https://docs.docker.com/engine/install/)
+* [Configure Docker for Cloud Container Registry](https://cloud.google.com/container-registry/docs/pushing-and-pulling)
+* [Upload the training data](https://cloud.google.com/storage/docs/uploading-objects) in the [TFRecord](https://www.tensorflow.org/tutorials/load_data/tfrecord) format to the storage bucket. You can preprocess your audio files into this format using the `ddsp_prepare_tfrecord` tool as described in [Making a TFRecord dataset from your own sounds](https://github.com/magenta/ddsp/tree/master/ddsp/training/data_preparation).
+
+### Quickstart:
+
+#### Define some environment variables
+
+If you are located outside Europe we recommend setting `$REGION` accordingly to your location. We also recommend to [setup hostname](https://cloud.google.com/container-registry/docs/pushing-and-pulling#tag_the_local_image_with_the_registry_name) in `$IMAGE_URI` based of the `$REGION` choice as if your Docker images are stored in different region than the job is computed additional charges will be applied.
```
Comment (marmolejo) [inline magenta_docker/README.md]: I think this readme should be region agnostic. In any case use the default us-central1 default region location just for explainability.

File: magenta_docker/config_multiple_vms.yaml [hunk 6]
```diff
@@ -0,0 +1,21 @@
+# Recommended batch_size: 128
+# Recommended learning_rate: 0.001
+# Recommended number of steps: 15000
```
Comment (marmolejo) [inline magenta_docker/config_multiple_vms.yaml:3]: You can add in the comment the link to the design doc explaining how you got this results.

File: magenta_docker/config_multiple_vms.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  # Configures 3 workers, each with 2 T4 GPUs
+  workerCount: 3
+  workerType: n1-highcpu-32
+  workerConfig:
+    acceleratorConfig:
+      count: 2
+      type: NVIDIA_TESLA_T4
+  # Makes AI Platform naming compatibile with Tensorflow naming
+  useChiefInTfConfig: True
+
```
Comment (marmolejo) [inline magenta_docker/config_multiple_vms.yaml]: (Here and below) Remove extra whitespace

File: magenta_docker/task_test.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
+"""Tests for task.py."""
```
Comment (marmolejo) [inline magenta_docker/task_test.py:1]: This is great test!

**Review by marmolejo — APPROVED** [review]

File: magenta_docker/config_multiple_vms.yaml (hunk 6 above)
Comment (werkaaa) [reply magenta_docker/config_multiple_vms.yaml:3]: Even for the PR to the Magenta library? It won't be externally visible

File: magenta_docker/Dockerfile (hunk 1 above)
Comment (werkaaa) [reply magenta_docker/Dockerfile]: Sorry, I'm not sure if I understand exactly what you mean. Should I move workdir definition to the place where I copy the script?

Comment (marmolejo) [reply magenta_docker/Dockerfile]: I mean the WORKDIR is used in the unzip below to create temporary files, you can move it before that.

Comment (werkaaa) [reply magenta_docker/Dockerfile]: Sure, will do

Comment (werkaaa) [reply magenta_docker/Dockerfile]: Done

File: magenta_docker/Dockerfile (hunk 2 above)
Comment (werkaaa) [reply magenta_docker/Dockerfile]: Done

File: magenta_docker/Dockerfile (hunk 3 above)
Comment (werkaaa) [reply magenta_docker/Dockerfile]: Done

File: magenta_docker/Dockerfile (hunk 4 above)
Comment (werkaaa) [reply magenta_docker/Dockerfile]: Done

File: magenta_docker/README.md (hunk 5 above)
Comment (werkaaa) [reply magenta_docker/README.md]: Done

File: magenta_docker/config_multiple_vms.yaml (hunk 7 above)
Comment (werkaaa) [reply magenta_docker/config_multiple_vms.yaml]: Done

File: magenta_docker/task_test.py (hunk 8 above)
Comment (werkaaa) [reply magenta_docker/task_test.py:1]: Thank you! :)

File: magenta_docker/config_multiple_vms.yaml (hunk 6 above)
Comment (werkaaa) [reply magenta_docker/config_multiple_vms.yaml:3]: I will add the Markdown files inside our repo when the design doc is in final stage.

**Pull request merged**